File: web/cockpit/shift-light-modos.js
// shift-light-modos.js — definição canônica dos 3 modos do shift light.
//
// Decisão Flávio 2026-05-29 após auditoria de engenheiro de competição sênior:
// Janelas RECALIBRADAS pra curva real do Bubi (não chutadas).
//
// Bubi: pico HP 6.050 / pico torque 5.200 / redline 6.300 / limite 6.350.
// Auditor: "Conservador 5.700-6.000 trocava ANTES do pico de potência — não
// era conservador, era desperdiçar HP". Bandas reescritas conforme recomendação.
//
// Esses 3 modos formam o ENVELOPE que a IA respeita ao escolher o ponto
// exato de troca em cada trecho da pista. IA opera SOMENTE dentro da janela.

export const ModoStint = Object.freeze({
  DURABILIDADE: 'durabilidade',
  NORMAL:       'normal',
  AGRESSIVO:    'agressivo',
});

/**
 * Perfil de motor — 4 pontos chave que definem a "personalidade" do motor.
 * Cada carro tem o seu, derivado da curva do dinamômetro.
 *
 * Decisão Flávio 2026-05-29: "dentro dos padrões de cada motor".
 * As 3 janelas (Durabilidade/Normal/Agressivo) são DERIVADAS desses 4 pontos —
 * não há janela hardcoded universal. Cada motor tem janelas próprias.
 */
export const PERFIL_BUBI = Object.freeze({
  apelido:             'Bubi (Celta 1.4 motor Onix)',
  picoTorqueRpm:       5200,  // Lenza 2026-05-18
  picoPotenciaRpm:     6050,
  redlineRpm:          6300,  // operacional
  limiteHardRpm:       6350,  // mecânico (acima disso, risco real de quebra)
});

/**
 * Deriva as 3 janelas dos modos a partir do perfil do motor.
 * Heurística (calibrada pra carro turismo amador, validada pelo auditor):
 *   - Durabilidade: começa pouco antes do pico de potência, termina no pico.
 *     Troca cedo, motor dura mais. Margem confortável.
 *   - Normal: começa logo após o pico de potência, vai até pouco antes do redline.
 *     Padrão recomendado.
 *   - Agressivo: começa antes do redline, vai até o limite mecânico.
 *     Aceita risco mecânico, máximo ganho de volta.
 *
 * Pra Bubi: pico HP 6050, redline 6300, limite 6350 →
 *   Durabilidade 5800-6000, Normal 6100-6250, Agressivo 6250-6350.
 *
 * Pra outro motor (ex: Civic 1.6 com pico HP em 7000, redline 7500),
 * a função gera janelas diferentes, calibradas pra esse motor.
 */
export function derivarJanelas(perfil) {
  const { picoPotenciaRpm, redlineRpm, limiteHardRpm } = perfil;
  return {
    [ModoStint.DURABILIDADE]: {
      nome:       'Durabilidade',
      nomeBreve:  'DUR',
      rpmMin:     picoPotenciaRpm - 250,
      rpmMax:     picoPotenciaRpm,
      descCurta:  'Stints longos, motor frio, chuva, treino sem pressa',
      descLonga:  'Troca pouco antes do pico de potência. Perde um pouco de volta mas preserva o motor. Use em treino livre tranquilo, condições molhadas ou quando o motor precisa durar a temporada inteira.',
      cor:        'azul',
      ganhoVoltaEstimadoS: -0.4,
      desgasteRelativo:    'baixo',
    },
    [ModoStint.NORMAL]: {
      nome:       'Normal',
      nomeBreve:  'NRM',
      rpmMin:     picoPotenciaRpm + 50,
      rpmMax:     redlineRpm - 50,
      descCurta:  'Padrão recomendado — equilíbrio entre volta e durabilidade',
      descLonga:  'Troca após o pico de potência com margem do redline. Aproveita a banda útil do motor sem forçar o limite mecânico. Modo recomendado pra maioria das sessões.',
      cor:        'verde',
      ganhoVoltaEstimadoS: 0.0,
      desgasteRelativo:    'normal',
    },
    [ModoStint.AGRESSIVO]: {
      nome:       'Agressivo',
      nomeBreve:  'AGR',
      rpmMin:     redlineRpm - 50,
      rpmMax:     limiteHardRpm,
      descCurta:  'Volta de qualificação — aceita risco mecânico',
      descLonga:  'Troca próximo do limite mecânico. Máximo ganho de volta possível. Use em volta de qualificação ou momentos decisivos. NÃO use em motor frio (água < 80°C) — desgaste real.',
      cor:        'vermelho',
      ganhoVoltaEstimadoS: 0.4,
      desgasteRelativo:    'alto',
    },
  };
}

// Janelas pra Bubi (default exportado pra retrocompatibilidade).
// Derivadas do PERFIL_BUBI pra o sistema não ter números mágicos.
export const JANELAS_BUBI = Object.freeze(derivarJanelas(PERFIL_BUBI));

// Envelope de segurança DURO (auditor recomendação 3).
// IA opera autônoma DENTRO desses limites, NUNCA os viola.
export const ENVELOPE_DEFAULT_BUBI = Object.freeze({
  rpm_max_absoluto:        6300,  // teto duro independente do modo
  rpm_min_motor_celsius:   80,    // motor frio = rebaixa pra Normal automaticamente
  forca_lateral_max_g:     0.50,  // nunca troca acima disso (proteção do eixo)
  oleo_max_celsius:        115,   // acima disso, alarma + reduz pra Durabilidade
  agua_max_celsius:        95,    // acima disso, alarma + reduz pra Durabilidade
});

/**
 * Retorna a janela do modo escolhido pra o motor especificado.
 * @param {string} modo - ModoStint.*
 * @param {Object} [perfilMotor=PERFIL_BUBI] — perfil do motor; default = Bubi
 * @returns {{rpmMin: number, rpmMax: number, nome: string, ...}}
 */
export function janelaParaModo(modo, perfilMotor = PERFIL_BUBI) {
  const janelas = perfilMotor === PERFIL_BUBI ? JANELAS_BUBI : derivarJanelas(perfilMotor);
  const j = janelas[modo];
  if (!j) throw new Error(`shift-light-modos: modo "${modo}" desconhecido`);
  return j;
}

/**
 * Lista os 3 modos em ordem canônica (Durabilidade, Normal, Agressivo).
 */
export function listarModos() {
  return [
    JANELAS_BUBI[ModoStint.DURABILIDADE],
    JANELAS_BUBI[ModoStint.NORMAL],
    JANELAS_BUBI[ModoStint.AGRESSIVO],
  ];
}

/**
 * Valida se um RPM proposto está dentro da janela do modo.
 * @param {number} rpm
 * @param {string} modo
 */
export function rpmDentroDaJanela(rpm, modo) {
  const j = janelaParaModo(modo);
  return rpm >= j.rpmMin && rpm <= j.rpmMax;
}

File: windows/cockpit/P1Fast.Cockpit.Domain/ShiftLightModos.cs
// ShiftLightModos — os 3 modos do shift light (Durabilidade / Normal / Agressivo).
// Port fiel de web/cockpit/shift-light-modos.js (decisão Flávio 2026-05-29, janelas
// recalibradas pra curva real do Bubi por auditor de competição sênior).
//
// As 3 janelas são DERIVADAS do perfil do motor (4 pontos-chave da curva do dyno) —
// NÃO há janela hardcoded universal; cada motor tem as suas. A IA da luz de marcha
// opera SOMENTE dentro da janela do modo escolhido.

namespace P1Fast.Cockpit.Domain;

/// <summary>Os 3 modos canônicos, em ordem (Durabilidade &lt; Normal &lt; Agressivo).</summary>
public enum ModoStint { Durabilidade, Normal, Agressivo }

/// <summary>Perfil do motor: os 4 pontos-chave que definem sua "personalidade" (do dyno).</summary>
public sealed record PerfilMotor(
    string Apelido,
    double PicoTorqueRpm,
    double PicoPotenciaRpm,
    double RedlineRpm,     // operacional
    double LimiteHardRpm); // mecânico (acima disso, risco real de quebra)

/// <summary>Janela de um modo: faixa de RPM de troca + rótulos e custo/ganho.</summary>
public sealed record JanelaModo(
    ModoStint Modo,
    string Nome,
    string NomeBreve,
    double RpmMin,
    double RpmMax,
    string Cor,
    double GanhoVoltaEstimadoS,
    string DesgasteRelativo);

public static class ShiftLightModos
{
    /// <summary>Bubi (Celta 1.4 motor Onix) — pico torque 5.200, pico HP 6.050, redline 6.300, limite 6.350.</summary>
    public static readonly PerfilMotor PerfilBubi = new(
        Apelido: "Bubi (Celta 1.4 motor Onix)",
        PicoTorqueRpm: 5200,
        PicoPotenciaRpm: 6050,
        RedlineRpm: 6300,
        LimiteHardRpm: 6350);

    /// <summary>
    /// Deriva as 3 janelas do perfil do motor (heurística validada pelo auditor):
    ///  - Durabilidade: começa pouco antes do pico de potência, termina no pico (troca cedo, motor dura).
    ///  - Normal: logo após o pico até pouco antes do redline (padrão recomendado).
    ///  - Agressivo: antes do redline até o limite mecânico (aceita risco, máximo ganho).
    /// Pra Bubi: Dur 5800-6050, Normal 6100-6250, Agr 6250-6350.
    /// </summary>
    public static IReadOnlyDictionary<ModoStint, JanelaModo> DerivarJanelas(PerfilMotor perfil)
    {
        ArgumentNullException.ThrowIfNull(perfil);
        var pico = perfil.PicoPotenciaRpm;
        var redline = perfil.RedlineRpm;
        var limite = perfil.LimiteHardRpm;
        return new Dictionary<ModoStint, JanelaModo>
        {
            [ModoStint.Durabilidade] = new(ModoStint.Durabilidade, "Durabilidade", "DUR",
                RpmMin: pico - 250, RpmMax: pico, Cor: "azul", GanhoVoltaEstimadoS: -0.4, DesgasteRelativo: "baixo"),
            [ModoStint.Normal] = new(ModoStint.Normal, "Normal", "NRM",
                RpmMin: pico + 50, RpmMax: redline - 50, Cor: "verde", GanhoVoltaEstimadoS: 0.0, DesgasteRelativo: "normal"),
            [ModoStint.Agressivo] = new(ModoStint.Agressivo, "Agressivo", "AGR",
                RpmMin: redline - 50, RpmMax: limite, Cor: "vermelho", GanhoVoltaEstimadoS: 0.4, DesgasteRelativo: "alto"),
        };
    }

    /// <summary>Janelas do Bubi (default), derivadas do <see cref="PerfilBubi"/>.</summary>
    public static readonly IReadOnlyDictionary<ModoStint, JanelaModo> JanelasBubi = DerivarJanelas(PerfilBubi);

    /// <summary>Janela do modo pro motor dado (default = Bubi).</summary>
    public static JanelaModo JanelaParaModo(ModoStint modo, PerfilMotor? perfil = null)
    {
        var janelas = perfil is null || ReferenceEquals(perfil, PerfilBubi) ? JanelasBubi : DerivarJanelas(perfil);
        return janelas[modo];
    }

    /// <summary>Os 3 modos em ordem canônica.</summary>
    public static IReadOnlyList<JanelaModo> Listar(PerfilMotor? perfil = null)
    {
        var j = perfil is null ? JanelasBubi : DerivarJanelas(perfil);
        return new[] { j[ModoStint.Durabilidade], j[ModoStint.Normal], j[ModoStint.Agressivo] };
    }

    /// <summary>Um RPM está dentro da janela do modo?</summary>
    public static bool RpmDentroDaJanela(double rpm, ModoStint modo, PerfilMotor? perfil = null)
    {
        var jan = JanelaParaModo(modo, perfil);
        return rpm >= jan.RpmMin && rpm <= jan.RpmMax;
    }
}
